service/gcs/core/gcs/storage.go

```diff
@@ -32,51 +33,87 @@ const (
+	}
+	return nil
+}
+
+// getLayerMounts computes the mount specs for the scratch and layers.
+func (c *gcsCore) getLayerMounts(scratch string, layers []prot.Layer) (scratchMount *mountSpec, layerMounts []*mountSpec, err error) {
+	layerMounts = make([]*mountSpec, len(layers))
 	for i, layer := range layers {
-		deviceID := layer.Path
-		deviceName, err := c.deviceIDToName(deviceID)
+		deviceName, pmem, err := deviceIDToName(layer.Path)
 		if err != nil {
-			return "", nil, err
+			return nil, nil, err
+		}
+		options := []string{"noload"}
+		if pmem {
+			// PMEM devices support DAX and should use it
+			options = append(options, "dax")
+		}
+		layerMounts[i] = &mountSpec{
+			Source:     deviceName,
+			FileSystem: "ext4",
+			Flags:      syscall.MS_RDONLY,
+			Options:    options,
 		}
-		layerDevices[i] = deviceName
 	}
 	// An empty scratch value indicates no scratch space is to be attached.
-	if scratch == "" {
-		scratchDevice = ""
-	} else {
-		scratchDevice, err = c.deviceIDToName(scratch)
+	if scratch != "" {
+		scratchDevice, _, err := deviceIDToName(scratch)
 		if err != nil {
-			return "", nil, err
+			return nil, nil, err
+		}
+		scratchMount = &mountSpec{
+			Source:     scratchDevice,
+			FileSystem: "ext4",
 		}
 	}
 
-	return scratchDevice, layerDevices, nil
+	return scratchMount, layerMounts, nil
 }
 
-// getMappedVirtualDiskDevices uses the Lun values in the given disks to
-// retrieve their associated device names.
-func (c *gcsCore) getMappedVirtualDiskDevices(disks []prot.MappedVirtualDisk) ([]string, error) {
-	devices := make([]string, len(disks))
+// getMappedVirtualDiskMounts uses the Lun values in the given disks to
+// retrieve their associated mount spec.
+func (c *gcsCore) getMappedVirtualDiskMounts(disks []prot.MappedVirtualDisk) ([]*mountSpec, error) {
+	devices := make([]*mountSpec, len(disks))
 	for i, disk := range disks {
-		device, err := c.deviceIDToName(strconv.Itoa(int(disk.Lun)))
+		device, err := scsiLunToName(int(disk.Lun))
 		if err != nil {
 			return nil, errors.Wrapf(err, "failed to get device name for mapped virtual disk %s, lun %d", disk.ContainerPath, disk.Lun)
 		}
-		devices[i] = device
+		flags := uintptr(0)
+		var options []string
+		if disk.ReadOnly {
+			flags |= syscall.MS_RDONLY
+			options = append(options, "noload")
+		}
+		devices[i] = &mountSpec{
+			Source:     device,
+			FileSystem: "ext4",
+			Flags:      flags,
+			Options:    options,
+		}
 	}
 	return devices, nil
 }
 
-// deviceIDToName converts from a SCSI location (0, 1, etc.) to a device (sda,
-// sdb, etc.).
-// NOTE: While this function currently works with SCSI codes, it will
-// eventually be reimplemented to use vSCM codes.
-func (c *gcsCore) deviceIDToName(deviceID string) (string, error) {
-	scsiID := fmt.Sprintf("0:0:0:%s", deviceID)
+// scsiLunToName finds the SCSI device with the given LUN. This assumes
+// only one SCSI controller.
+func scsiLunToName(lun int) (string, error) {
```

Comment:
What is the point of changing the type? This is `uint8` on `MappedVirtualDisk` before the cast and makes more sense that way. What does `0:0:0:-1` mean.